sys/sleepq.c

```diff
@@ -209,21 +214,37 @@ void sleepq_wait(void *wchan, const void *waitpt) {
       td->td_state = TDS_SLEEPING;
       sched_switch();
     }
+    /* TDF_SLPINTR flag is set after wakeup only if sleep was aborted. */
+    if (td->td_flags & TDF_SLPINTR) {
+      td->td_flags &= ~TDF_SLPINTR;
+      wakeup = SQ_ABORT;
+    }
   }
+
+  return wakeup;
 }
 
 /* Remove a thread from the sleep queue and resume it. */
-static void sq_wakeup(thread_t *td, sleepq_chain_t *sc, sleepq_t *sq) {
+static bool sq_wakeup(thread_t *td, sleepq_chain_t *sc, sleepq_t *sq,
+                      sq_wakeup_t wakeup) {
+  /* Do not try to abort thread's sleep if it's not prepared for that. */
+  if (!(td->td_flags & TDF_SLPINTR) && (wakeup == SQ_ABORT))
+    return false;
+
   sq_leave(td, sc, sq);
 
   WITH_SPINLOCK(td->td_spin) {
+    /* Clear TDF_SLPINTR flag if thread's sleep was not aborted. */
+    if (wakeup == SQ_NORMAL)
+      td->td_flags &= ~TDF_SLPINTR;
     /* Do not try to wake up a thread that is sleepy but did not fall asleep! */
     if (td->td_flags & TDF_SLEEPY) {
       td->td_flags &= ~TDF_SLEEPY;
-    } else {
-      sched_wakeup(td, 0);
+      return false;
```

Comment:
I think we should consider SLEEPY thread to be sleeping for purposes of this function. I mean, `sq_wakeup` should return `true` if the thread was actually on some sleepq and accepted abort. SLEEPY thread is indeed on some sleepq.

Reply:
Ok. Done.